src/primitives/iter.rs

```diff
@@ -0,0 +1,501 @@
+    /// checksum engine of the iterator.
+    pub fn with_checksummed_fe(mut self, fe: Fe32) -> Self {
+        self.checksum_engine.input_fe(fe);
+        self
+    }
+}
+
+impl<I, Ck> Iterator for ChecksumIter<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    type Item = Fe32;
+
+    fn next(&mut self) -> Option<Fe32> {
+        match self.iter.next() {
+            Some(fe) => {
+                self.checksum_engine.input_fe(fe);
+                Some(fe)
+            }
+            None =>
+                if self.checksum_remaining == 0 {
+                    None
+                } else {
+                    if self.checksum_remaining == Ck::CHECKSUM_LENGTH {
+                        self.checksum_engine.input_target_residue();
+                    }
+                    self.checksum_remaining -= 1;
+                    Some(Fe32(self.checksum_engine.residue().unpack(self.checksum_remaining)))
+                },
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.iter.size_hint();
+        (min + self.checksum_remaining, max.map(|max| max + self.checksum_remaining))
+    }
+}
+
+/// Iterator adaptor which takes a stream of field elements, converts it to characters prefixed by
+/// an HRP. If `fe_iter` is a checksummed iter, it is expected that the `hrp` strings are identical.
+pub struct HrpCharIter<'hrp, I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    hrp_iter: hrp::LowercaseCharIter<'hrp>,
+    fe_iter: I,
+    hrp_done: bool,
+}
+
+impl<'hrp, I> Iterator for HrpCharIter<'hrp, I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    type Item = char;
+
+    fn next(&mut self) -> Option<char> {
+        if !self.hrp_done {
+            match self.hrp_iter.next() {
+                Some(c) => return Some(c),
+                None => {
+                    self.hrp_done = true;
+                    return Some('1');
+                }
+            }
+        }
+        self.fe_iter.next().map(Fe32::to_char)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.fe_iter.size_hint();
+        let add = if !self.hrp_done {
+            self.hrp_iter.len() + 1 // hrp | SEP
+        } else {
+            0
+        };
+
+        (min + add, max.map(|max| max + add))
+    }
+}
+
+/// Returns the number of Fe32's encoded by `n` bytes.
+fn bytes_len_to_fes_len(n: usize) -> usize {
+    let bits = n * 8;
+    if bits % 5 != 0 {
+        bits / 5 + 1 // +1 because we pad the fes.
+    } else {
+        bits / 5
+    }
+}
+
+/// Returns the number of bytes required to encode `n` Fe32's.
+///
+/// (If the total number of bits is not a multiple of 8, any trailing bits are dropped.)
+fn fes_len_to_bytes_len(n: usize) -> usize { n * 5 / 8 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    /// Asserts two iterators are equal.
+    fn check_iter_eq<I, J, T>(mut i: I, mut j: J)
```

Comment:
In 0c6613106fa624db037598dab5e9a19634acd2d0:

Could this whole function be replaced with [Iterator::eq](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.eq)?

Reply:
yep, seems so. Done.